rust/worker/src/segment/metadata_segment.rs

```diff
@@ -418,7 +481,210 @@ impl<'a> SegmentWriter<'a> for MetadataSegmentWriter {
+                                        MetadataValue::Float(value) => {
+                                            match &self.f32_metadata_index_writer {
+                                                Some(writer) => {
+                                                    let _ = writer.delete(
+                                                        key,
+                                                        *value as f32,
+                                                        segment_offset_id,
+                                                    );
+                                                }
+                                                None => {}
+                                            }
+                                        }
+                                        MetadataValue::Int(value) => {
+                                            match &self.u32_metadata_index_writer {
+                                                Some(writer) => {
+                                                    let _ = writer.delete(
+                                                        key,
+                                                        *value as u32,
+                                                        segment_offset_id,
+                                                    );
+                                                }
+                                                None => {}
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                            None => {}
+                        };
+                        match &data_record.document {
+                            Some(document) => match &self.full_text_index_writer {
+                                Some(writer) => {
+                                    let _ = writer.delete_document(document, segment_offset_id);
+                                }
+                                None => {}
+                            },
+                            None => {}
+                        };
                     }
+                    None => {}
+                },
+                Operation::Update => {
+                    // TODO figure out how to not do this without adding another
+                    // layer of `match`ing.
+                    let empty_hashmap = HashMap::new();
+                    let old_metadata = match &record.0.data_record {
+                        Some(data_record) => match &data_record.metadata {
+                            Some(metadata) => metadata,
+                            None => &empty_hashmap,
+                        },
+                        None => &empty_hashmap,
+                    };
+                    match &record.0.data_record {
+                        Some(data_record) => match &data_record.metadata {
```

Comment:
Trying to understand this logic - aren't both old and new values from the same `record.0.data_record` here? Shouldn't new value be from `record.0.metadata_to_be_merged`?

Reply:
Good point. So I need to make sure the data_record exists, then check if metadata_to_be_merged exists and then do the merge, right?

Reply:
Resolved offline